source/server/server.cc

```diff
@@ -387,25 +387,44 @@ void InstanceImpl::loadServerFlags(const absl::optional<std::string>& flags_path
 
 uint64_t InstanceImpl::numConnections() { return listener_manager_->numConnections(); }
 
-RunHelper::RunHelper(Event::Dispatcher& dispatcher, Upstream::ClusterManager& cm,
+RunHelper::RunHelper(Options& options, Event::Dispatcher& dispatcher, Upstream::ClusterManager& cm,
                      HotRestart& hot_restart, AccessLog::AccessLogManager& access_log_manager,
                      InitManagerImpl& init_manager, OverloadManager& overload_manager,
                      std::function<void()> workers_start_cb) {
 
   // Setup signals.
-  sigterm_ = dispatcher.listenForSignal(SIGTERM, [this, &hot_restart, &dispatcher]() {
-    ENVOY_LOG(warn, "caught SIGTERM");
-    shutdown(dispatcher, hot_restart);
-  });
+  if (options.signalHandler().allows("SIGTERM")) {
+    sigterm_ = dispatcher.listenForSignal(SIGTERM, [this, &hot_restart, &dispatcher]() {
+      ENVOY_LOG(warn, "caught SIGTERM");
+      shutdown(dispatcher, hot_restart);
+    });
+  } else {
+    sigterm_ = dispatcher.listenForSignal(SIGTERM, [this]() {
```

Comment:
Why are you listening for sigterm when told not to? I think we don't want to even register the listener, which might conflict with other listeners that are linked into the Envoy binary.